### PR TITLE
Restoring threaded QML rendering

### PR DESCRIPTION
**Pull request by jherico**
## Testing

Application behavior should be unchanged.   Test UI in 2D and HMD mode.  Verify that web entities and web overlays still work

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 263be6efdcba3f8bb2b4417157d9f23266732a83 available at: https://gist.github.com/6c0139d7beea9418f4a566709605552d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-263be6e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-263be6e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 04d343dc60806e38c6a98ebd31d58386496e23a8 available at: https://gist.github.com/6c0139d7beea9418f4a566709605552d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-04d343d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-04d343d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dfd70df3ef0c9fd199602e566ac853e936770ee8 available at: https://gist.github.com/6c0139d7beea9418f4a566709605552d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-dfd70df.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-dfd70df.dmg

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 1]
```diff
@@ -253,14 +250,202 @@ QNetworkAccessManager* QmlNetworkAccessManagerFactory::create(QObject* parent) {
...
+    // Event-driven methods
+    void init();
+    void render();
+    void resize();
+    void cleanup();
+
+    // Helper methods
+    void setupFbo();
+
+    //bool allowNewFrame(uint8_t fps);
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: remove dead code?

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 2]
```diff
@@ -273,50 +458,106 @@ void OffscreenQmlSurface::cleanup() {
...
+        GLsync fence = glFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
+        // Fence will be queried from another context, so we MUST flush here
         glFlush();
+        _surface->updateTexture(OffscreenQmlSurface::TextureAndFence { texture, fence });
+        _quickWindow->resetOpenGLState();
+        _lastRenderTime = usecTimestampNow();
+    } catch (std::runtime_error& error) {
+        // Don't leak a texture
+        offscreenTextures.releaseTexture(OffscreenQmlSurface::TextureAndFence { texture, 0 });
+        qWarning() << "Failed to render QML: " << error.what();
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: use `qCWarning` and a category

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 3]
```diff
@@ -349,17 +590,23 @@ bool OffscreenQmlSurface::allowNewFrame(uint8_t fps) {
...
+static const uint64_t MAX_SHUTDOWN_WAIT_SECS = 2;
 OffscreenQmlSurface::~OffscreenQmlSurface() {
     QObject::disconnect(&_updateTimer);
     QObject::disconnect(qApp);
 
-    cleanup();
+    qCDebug(glLogging) << "Stopping QML Renderer Thread " << _renderer->currentThreadId();
+    _renderer->_queue.add(STOP);
+    if (!_renderer->wait(MAX_SHUTDOWN_WAIT_SECS * USECS_PER_SECOND)) {
+        qWarning() << "Failed to shut down the QML Renderer Thread";
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: `qCWarning`

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 4]
```diff
@@ -349,17 +590,23 @@ bool OffscreenQmlSurface::allowNewFrame(uint8_t fps) {
...
+    }
+//    cleanup();
 
     _canvas->deleteLater();
     _rootItem->deleteLater();
     _qmlComponent->deleteLater();
     _qmlEngine->deleteLater();
-    _quickWindow->deleteLater();
+//    _quickWindow->deleteLater();
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: dead code?

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 5]
```diff
@@ -349,17 +590,23 @@ bool OffscreenQmlSurface::allowNewFrame(uint8_t fps) {
...
     QObject::disconnect(&_updateTimer);
     QObject::disconnect(qApp);
 
-    cleanup();
+    qCDebug(glLogging) << "Stopping QML Renderer Thread " << _renderer->currentThreadId();
+    _renderer->_queue.add(STOP);
+    if (!_renderer->wait(MAX_SHUTDOWN_WAIT_SECS * USECS_PER_SECOND)) {
+        qWarning() << "Failed to shut down the QML Renderer Thread";
+    }
+//    cleanup();
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: dead code?

File: libraries/gl/src/gl/OffscreenQmlSurface.h [hunk 6]
```diff
@@ -113,21 +117,22 @@ public slots:
...
 
 private:
+    void updateTexture(const TextureAndFence& textureAndFence);
     QObject* finishQmlLoad(std::function<void(QQmlContext*, QObject*)> f);
     QPointF mapWindowToUi(const QPointF& sourcePosition, QObject* sourceObject);
     void setupFbo();
     bool allowNewFrame(uint8_t fps);
     void render();
-    void cleanup();
+    //void cleanup();
```
Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.h]: dead code

**Comment by hifi-gustavo** [comment]
Build console output for commit: 697e7233c7a67bd39c5fc3cc597d1077b1fca69e available at: https://gist.github.com/6c0139d7beea9418f4a566709605552d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-697e723.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9292-697e723.dmg

**Comment by claudioshg** [comment]
### Tested Windows 10 - Oculus CV1

Verify that web entities and web overlays still work in 2D and in HMD:

Scrolling up and down is **very buggy**. Sometimes responsive and some not so much.
I was unable to link to any sites featured on the front page. 
At the bottom of the front page there are links to High Fidelity's Youtube, Worklist, Git-Hub, Twitter, Ect... they are unresponsive. I was unable to navigate to pages that required opening a new tab, I believe.
I was able to navigate to STEAM and one other page. Some links worked and some didn't from within the WEB entity.
I was able to navigate the internet by the means of CTRL+B.

**Pull request closed without merging**